zerver/lib/test_classes.py

```diff
@@ -850,3 +854,68 @@ def setUp(self) -> None:
 
     def setUpBeforeMigration(self, apps: StateApps) -> None:
         pass  # nocoverage
+
+class LDAPTestCase(ZulipTestCase):
+    """
+    Base class for LDAP tests. Takes care of the mock_ldap setup, loads
+    a directory from fixtures/ldap/directory.json with various entries to be
+    used by tests. If a test wants to specify its own directory, it can just replace
+    self.mock_ldap.directory with its own content, but in most cases it should be
+    enough to use change_user_attr to make simple modifications to the pre-loaded
+    directory. If new user entries are needed to test for some additional unusual
+    scenario, it's most likely best to add that to directory.json.
+    """
+
+    def setUp(self) -> None:
+        directory = ujson.loads(self.fixture_data("directory.json", type="ldap"))
+
+        # Load binary attributes. If in "directory", an attribute as its value
+        # has a string starting with "file:", the rest of the string is assumed
+        # to be a path to the file from which binary data should be loaded,
+        # as the actual value of the attribute in ldap.
+        for dn, attrs in directory.items():
+            for attr, value in attrs.items():
+                if isinstance(value, str) and value.startswith("file:"):
+                    with open(value[5:], 'rb') as f:
+                        attrs[attr] = [f.read(), ]
+
+        ldap_patcher = mock.patch('django_auth_ldap.config.ldap.initialize')
+        self.mock_initialize = ldap_patcher.start()
+        self.mock_ldap = MockLDAP(directory)
+        self.mock_initialize.return_value = self.mock_ldap
+
+    def tearDown(self) -> None:
+        self.mock_ldap.reset()
+        self.mock_initialize.stop()
+
+    def settings(self, **kwargs: Any) -> Any:
+        """
+        Extend self.settings for LDAP tests to set various LDAP defaults if not explicitly specified:
+        The server URI, and how to find users by username or email.
+        """
+        if "AUTH_LDAP_USER_SEARCH" not in kwargs:
+            kwargs["AUTH_LDAP_USER_SEARCH"] = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                                         ldap.SCOPE_ONELEVEL, "(uid=%(user)s)")
+        settings_kwargs = {}  # type: Dict[str, Any]
+        for key, value in kwargs.items():
+            settings_kwargs[key] = value
+        return super().settings(**settings_kwargs)
+
+    def change_user_attr(self, username: str, attr_name: str, attr_value: Union[str, bytes],
```

Comment:
Let's call this `change_ldap_user_attr`.

Reply:
Changed (this is now a method of ZulipTestCase).